module/VuFind/src/VuFind/Db/Service/AuditEventService.php

```diff
@@ -0,0 +1,280 @@
+     *
+     * @return void
+     */
+    public function addEvent(
+        AuditEventType $type,
+        AuditEventSubtype $subtype,
+        ?UserEntityInterface $user = null,
+        ?string $message = null,
+        array $data = []
+    ): void {
+        if (!in_array($type->value, $this->enabledEventTypes)) {
+            return;
+        }
+        $data = $this->scrubSecrets($data);
+        $data['__method'] = $this->getCallerOfParentMethod();
+        $event = $this->createEntity();
+        $event
+            ->setType($type)
+            ->setSubtype($subtype)
+            ->setUser($user)
+            ->setSessionId('cli' !== PHP_SAPI ? $this->sessionManager->getId() : null)
+            ->setClientIp($_SERVER['REMOTE_ADDR'] ?? null)
+            ->setServerIp($_SERVER['SERVER_ADDR'] ?? null)
+            ->setServerName($_SERVER['SERVER_NAME'] ?? null)
+            ->setMessage($message)
+            ->setData(json_encode($data));
+        $this->persistEntity($event);
+    }
+
+    /**
+     * Get an array of events.
+     *
+     * @param ?DateTime                    $fromDate   Start date
+     * @param ?DateTime                    $untilDate  End date
+     * @param ?string                      $type       Event type
+     * @param ?string                      $subtype    Event subtype
+     * @param UserEntityInterface|int|null $userOrId   User entity or ID of user
+     * @param ?string                      $username   User's username
+     * @param ?string                      $clientIp   Client's IP address
+     * @param ?string                      $serverIp   Server's IP address
+     * @param ?string                      $serverName Server's host name
+     * @param array                        $sort       Sort order
+     *
+     * @return AuditEventEntityInterface[]
+     */
+    public function getEvents(
+        ?DateTime $fromDate = null,
+        ?DateTime $untilDate = null,
+        ?string $type = null,
+        ?string $subtype = null,
+        UserEntityInterface|int|null $userOrId = null,
+        ?string $username = null,
+        ?string $clientIp = null,
+        ?string $serverIp = null,
+        ?string $serverName = null,
+        array $sort = ['date DESC']
+    ): array {
+        $user = $userOrId instanceof UserEntityInterface ? $userOrId->getId() : $userOrId;
+
+        $dql = 'SELECT e FROM ' . AuditEventEntityInterface::class . ' e';
+        $conditions = [];
+        $params = [];
+
+        // Handle date limits:
+        if (null !== $fromDate) {
+            $conditions[] = 'e.date >= :fromDate';
+            $params['fromDate'] = $fromDate->format('Y-m-d H:i:s');
+        }
+        if (null !== $untilDate) {
+            $conditions[] = 'e.date <= :untilDate';
+            $params['untilDate'] = $untilDate->format('Y-m-d H:i:s');
+        }
+        if (null !== $type) {
+            $conditions[] = 'e.type = :type';
+            $params[$type] = $type;
```

Comment:
Wrong key formatting -- and the error is repeated in several more conditions below.

```suggestion
            $params['type'] = $type;
```

Reply:
Oops. Fixed.